### PR TITLE
Enable OpaquePointers on LLVM 15

### PR DESCRIPTION
**Pull request by vchuravy**

### CODE REVIEW

**Comment by vchuravy** [comment]
@nanosoldier `runbenchmarks(!"scalar", vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/49846#issuecomment-1550333897) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/3868ae0_vs_c245179/report.md).

**Comment by maleadt** [comment]
Opaque pointers are a per-context setting, right? It would be nice if the GPU stack can keep working with opaque pointers disabled, as we have a couple of (closed-source) back-ends that haven't been adapted yet.

**Comment by Seelengrab** [comment]
If this becomes a flag in GPUCompiler, it would be great to have some API for setting that as well.

**Comment by maleadt** [comment]
LLVM.jl has a wrapper for the LLVM C API already: https://github.com/maleadt/LLVM.jl/blob/e44c430c4a48225114f217c36b166b85a3fcbe5a/lib/15/libLLVM_h.jl#L602-L604. Now that GPUCompiler create its own contexts for use with Julia, theoretically this should just work, but we should verify.

**Comment by vtjnash** [comment]
@nanosoldier `runbenchmarks("union", vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/49846#issuecomment-1769128809) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/6323bac_vs_3cc4fdc/report.md).

**Comment by vchuravy** [comment]
@pxl-th @maleadt do you think the GPU ecosystem is at a point where turning opaque-pointer should be fine?

@vtjnash and I are inclined on giving this a go.

**Comment by pxl-th** [comment]
Last time I tried opaque pointers it needed some adjustments on GPUCompiler side:
https://github.com/JuliaGPU/GPUCompiler.jl/issues/511.

**Comment by vtjnash** [comment]
I changed this PR so it only turns them on for the JIT and not globally.

**Comment by maleadt** [comment]
The main problem isn't GPUCompiler, it's the out-of-tree back-ends (SPIR-V translator for oneAPI, the AIR "back-end" for Metal, and NVVM for CUDA). Updating these is going to be a pain regardless, so it shouldn't hold up the migration here. I just was hoping that the next LTS would still support typed pointers, so that we have something to point users to while we're updating back-ends.

**Comment by maleadt** [comment]
I've been testing this, and it works great. I wonder if we could have the NFC version of this in 1.10, i.e., calling `setOpaquePointers` instead of using the global option. That way, we can start migrating GPUCompiler even if Julia is still using typed pointers (currently, using the global option "taints" the context, making it impossible to change the setting afterwards).

File: src/jitlayers.cpp [hunk 1]
```diff
@@ -1629,6 +1629,10 @@ JuliaOJIT::JuliaOJIT()
     DLSymOpt(std::make_unique<DLSymOptimizer>(false)),
     ContextPool([](){
         auto ctx = std::make_unique<LLVMContext>();
+#ifndef JL_LLVM_OPAQUE_POINTERS
+        if (ctx.supportsTypedPointers())
+            ctx.setOpaquePointers(false);
```
Comment (maleadt) [inline src/jitlayers.cpp:1634]:
```suggestion
        if (ctx->supportsTypedPointers())
            ctx->setOpaquePointers(false);
```

I've made that change in https://github.com/JuliaLang/julia/pull/51840

**Comment by maleadt** [comment]
For the record, with https://github.com/JuliaLang/julia/pull/51840 it doesn't matter whether Julia uses opaque pointers or not (LLVM's default is still typed, and Julia doesn't use the clopt anymore allowing GPUCompiler to override per-context). But I've also started work on updating GPUCompiler to opaque pointers in https://github.com/JuliaGPU/GPUCompiler.jl/pull/528. Regardless, it should be fine to flip the switch here.

Let's see what else breaks though:

@nanosoldier `runtests()`

**Comment by nanosoldier** [comment]
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/49846#issuecomment-1781820807) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/7e89102_vs_3493473/report.html) is available.

**Comment by vchuravy** [comment]
Will fix #52363 on for 1.11-dev

**Comment by gbaraldi** [comment]
Opaque pointers introduce a bunch of regressions that are only fixed on a patched llvm 16

**Comment by vtjnash** [comment]
Did it get worse? The results above seemed to look fine
@nanosoldier `runbenchmarks(!"scalar", vs=":master")`

**Comment by gbaraldi** [comment]
Unions are on average 50% slower

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/49846#issuecomment-1852755265) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/548b7ba_vs_eba10dd/report.md).

**Comment by vchuravy** [comment]
Superseded by #51720

**Pull request closed without merging**